src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+                next = i + 1;
+            }
+
+            _buckets = newBuckets;
+            _entries = newEntries;
+        }
+
+        private int IndexOf(TKey key, out int hashCode)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            IEqualityComparer<TKey> comparer = _comparer;
+            hashCode = (comparer?.GetHashCode(key) ?? key.GetHashCode()) & 0x7FFFFFFF;
+            int[] buckets = _buckets;
+            int index = -1;
+            if (buckets.Length > 0)
+            {
+                int bucket = hashCode % buckets.Length;
+                index = buckets[bucket] - 1;
+                comparer = comparer ?? EqualityComparer<TKey>.Default;
+                Entry[] entries = _entries;
+                while (index >= 0)
+                {
+                    Entry entry = entries[index];
+                    if (entry.HashCode == hashCode && comparer.Equals(entry.Key, key))
+                    {
+                        break;
+                    }
+                    index = entry.Next;
+                }
+            }
+            return index;
+        }
+
+        private bool TryInsert(int? index, TKey key, TValue value, InsertionBehavior behavior)
+        {
+            int i = IndexOf(key, out int hashCode);
+            Entry[] entries = _entries;
+            ++_version;
+            if (i >= 0)
+            {
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        entries[i].Value = value;
+                        return true;
+                    case InsertionBehavior.ThrowOnExisting:
+                        throw new ArgumentException("An item with the same key has already been added.");
+                    default:
+                        return false;
+                }
+            }
+
+            // Check if resize is needed
+            int count = Count;
+            if (entries.Length == count)
+            {
+                Resize(HashHelpers.ExpandPrime(entries.Length));
+                entries = _entries;
+            }
+
+            // Increment indices >= index;
+            int actualIndex = index ?? count;
+            for (i = actualIndex; i < count; ++i)
+            {
+                UpdateBucket(i, false, 1);
+            }
+            Array.Copy(entries, actualIndex, entries, actualIndex + 1, count - actualIndex);
+
+            Debug.Assert(_buckets.Length > 0);
+            ref int bucket = ref _buckets[hashCode % _buckets.Length];
+            Entry newEntry = new Entry { HashCode = hashCode, Key = key, Value = value, Next = bucket - 1 };
+            entries[actualIndex] = newEntry;
+            bucket = actualIndex + 1;
+            ++Count;
+            return true;
+        }
+
+        private void UpdateBucket(int i, bool setToNext, int incrementAmount = 0)
```

Comment:
This method is hard to read because it is overloaded for two purposes: removing an entry from the bucket, and re-numbering the index in the bucket. I think it would be clearer to have two methods. It might even be faster (fewer branches)